File: scripts/ingest-pdf-knowledge.ts
/**
 * scripts/ingest-pdf-knowledge.ts
 *
 * Chunks and embeds the NON-CRITERIA sections of the full KMS PDF
 * (everything EXCEPT pages 75–253, which are already in cia_criteria).
 *
 * What gets indexed:
 *   Pages   4– 74 : Dasar Teori (Puzzle Kesuksesan, fondasi KMS)
 *   Pages 254–258 : Cara Menanamkan KMS Quranik (5 langkah)
 *   Pages 259–596 : 25 Simulasi Situasi (kegagalan, konflik, dll.)
 *
 * REQUIREMENTS:
 *   1. Install the PDF parser (one-time):
 *        npm install --save-dev pdf-parse @types/pdf-parse
 *   2. Make sure OPENROUTER_API_KEY and NEXT_PUBLIC_SUPABASE_URL are in .env.local
 *      (SUPABASE_SERVICE_ROLE_KEY is preferred; anon key also works for inserts
 *       if RLS allows it)
 *   3. Run the SQL migration first:
 *        Supabase → SQL Editor → paste scripts/add_pdf_knowledge.sql → Run
 *   4. Run this script:
 *        npx tsx scripts/ingest-pdf-knowledge.ts --pdf /path/to/full-597-page.pdf
 *
 * Options:
 *   --pdf       <path>   Path to the full PDF  [REQUIRED]
 *   --skip-from <n>      First page to skip    [default: 75]
 *   --skip-to   <n>      Last page to skip     [default: 253]
 *   --chunk-words <n>    Target words per chunk [default: 300]
 *   --clear              Delete existing rows before ingesting (re-ingest)
 */

import * as fs from "fs";
import * as path from "path";
import { createClient } from "@supabase/supabase-js";
import OpenAI from "openai";
import dotenv from "dotenv";
import pdfParse from "pdf-parse";

dotenv.config({ path: ".env.local" });

// ─── Config ──────────────────────────────────────────────────────────────────

const OPENROUTER_API_KEY = process.env.OPENROUTER_API_KEY;
const SUPABASE_URL       = process.env.NEXT_PUBLIC_SUPABASE_URL;
const SUPABASE_KEY       =
  process.env.SUPABASE_SERVICE_ROLE_KEY ?? process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;

if (!OPENROUTER_API_KEY) throw new Error("❌  OPENROUTER_API_KEY missing in .env.local");
if (!SUPABASE_URL || !SUPABASE_KEY) throw new Error("❌  Supabase credentials missing in .env.local");

const openai = new OpenAI({
  baseURL: "https://openrouter.ai/api/v1",
  apiKey: OPENROUTER_API_KEY,
  defaultHeaders: { "X-Title": "CIA Assessment PDF Ingest" },
});

const supabase = createClient(SUPABASE_URL, SUPABASE_KEY);

// ─── Section map (derived from the PDF table of contents) ────────────────────
// Each entry defines a named section and its page range.
// Pages outside these ranges (and outside the skip range) get a fallback label.

const SECTIONS: { start: number; end: number; name: string }[] = [
  { start: 4,   end: 14,  name: "Puzzle Kesuksesan" },
  { start: 15,  end: 74,  name: "Dasar Teori QCB QMB QSB" },
  // 75–253: SKIP — already in cia_criteria
  { start: 254, end: 258, name: "Cara Menanamkan KMS Quranik" },
  { start: 259, end: 271, name: "Simulasi: Berpisah dengan Orangtua" },
  { start: 272, end: 284, name: "Simulasi: Tekanan Akademik" },
  { start: 285, end: 296, name: "Simulasi: Konflik Sosial" },
  { start: 297, end: 309, name: "Simulasi: Perubahan Lingkungan" },
  { start: 310, end: 321, name: "Simulasi: Kegagalan atau Penolakan" },
  { start: 322, end: 333, name: "Simulasi: Situasi Tidak Terduga" },
  { start: 334, end: 344, name: "Simulasi: Tekanan Moral atau Etika" },
  { start: 345, end: 356, name: "Simulasi: Medsos dan Perbandingan Diri" },
  { start: 357, end: 368, name: "Simulasi: Pengambilan Keputusan" },
  { start: 369, end: 380, name: "Simulasi: Kehilangan atau Duka" },
  { start: 381, end: 392, name: "Simulasi: Tantangan Identitas Diri" },
  { start: 393, end: 403, name: "Simulasi: Kompetisi dan Persaingan" },
  { start: 404, end: 419, name: "Simulasi: Kritik atau Teguran" },
  { start: 420, end: 436, name: "Simulasi: Penolakan" },
  { start: 437, end: 454, name: "Simulasi: Permintaan Bantuan" },
  { start: 455, end: 474, name: "Simulasi: Tidak Adil atau Diskriminatif" },
  { start: 475, end: 493, name: "Simulasi: Situasi Menyebalkan" },
  { start: 494, end: 512, name: "Simulasi: Dominan atau Terlalu Mengatur" },
  { start: 513, end: 528, name: "Simulasi: Pasif atau Tidak Kooperatif" },
  { start: 529, end: 542, name: "Simulasi: Meremehkan atau Merendahkan" },
  { start: 543, end: 558, name: "Simulasi: Emosional atau Marah" },
  { start: 559, end: 570, name: "Simulasi: Manipulatif atau Memperalat" },
  { start: 571, end: 587, name: "Simulasi: Tidak Konsisten" },
  { start: 588, end: 596, name: "Simulasi: Kompetitif Berlebihan" },
];

function getSectionName(pageNumber: number): string {
  const match = SECTIONS.find((s) => pageNumber >= s.start && pageNumber <= s.end);
  return match?.name ?? `Halaman ${pageNumber}`;
}

// ─── CLI argument parsing ─────────────────────────────────────────────────────

function getArg(flag: string, fallback?: string): string | undefined {
  const idx = process.argv.indexOf(flag);
  return idx !== -1 ? process.argv[idx + 1] : fallback;
}

const pdfPath    = getArg("--pdf");
const skipFrom   = parseInt(getArg("--skip-from", "75")!,  10);
const skipTo     = parseInt(getArg("--skip-to",   "253")!, 10);
const chunkWords = parseInt(getArg("--chunk-words","300")!, 10);
const clearFirst = process.argv.includes("--clear");

if (!pdfPath) {
  console.error("❌  Usage: npx tsx scripts/ingest-pdf-knowledge.ts --pdf /path/to/book.pdf");
  process.exit(1);
}

// ─── PDF extraction ───────────────────────────────────────────────────────────

interface PageText {
  pageNumber: number; // 1-based
  text: string;
}

async function extractPages(filePath: string): Promise<PageText[]> {
  const buffer = fs.readFileSync(path.resolve(filePath));
  const pages: PageText[] = [];

  // pdf-parse calls the render callback once per page
  await pdfParse(buffer, {
    pagerender: (pageData: any) => {
      return pageData.getTextContent().then((content: any) => {
        const text = content.items
          .map((item: any) => item.str)
          .join(" ")
          .replace(/\s+/g, " ")
          .trim();
        pages.push({ pageNumber: pageData.pageNumber, text });
        return text;
      });
    },
  });

  return pages;
}

// ─── Chunking ─────────────────────────────────────────────────────────────────
// Groups consecutive pages that share the same section name into ~chunkWords
// word chunks.  We split within a section rather than across sections so that
// every chunk stays topically coherent.

interface Chunk {
  content: string;
  section: string;
  pageStart: number;
  pageEnd: number;
}

function chunkPages(pages: PageText[]): Chunk[] {
  const chunks: Chunk[] = [];

  // Group pages by section
  const sectionGroups: Map<string, PageText[]> = new Map();
  const sectionOrder: string[] = [];

  for (const page of pages) {
    const section = getSectionName(page.pageNumber);
    if (!sectionGroups.has(section)) {
      sectionGroups.set(section, []);
      sectionOrder.push(section);
    }
    sectionGroups.get(section)!.push(page);
  }

  for (const section of sectionOrder) {
    const sectionPages = sectionGroups.get(section)!;
    const fullText = sectionPages.map((p) => p.text).join(" ").replace(/\s+/g, " ").trim();
    if (!fullText) continue;

    const words = fullText.split(" ");
    let start = 0;

    while (start < words.length) {
      const end = Math.min(start + chunkWords, words.length);
      const chunkText = words.slice(start, end).join(" ").trim();

      if (chunkText.length > 50) { // skip near-empty chunks
        chunks.push({
          content: `[${section}]\n${chunkText}`,
          section,
          pageStart: sectionPages[0].pageNumber,
          pageEnd:   sectionPages[sectionPages.length - 1].pageNumber,
        });
      }

      start = end;
    }
  }

  return chunks;
}

// ─── Embedding ────────────────────────────────────────────────────────────────

async function embedText(text: string): Promise<number[]> {
  const response = await openai.embeddings.create({
    model: "openai/text-embedding-3-small",
    input: text,
  });
  return response.data[0].embedding;
}

// ─── Main ─────────────────────────────────────────────────────────────────────

async function main() {
  console.log(`\n📖  Reading PDF: ${pdfPath}`);
  const allPages = await extractPages(pdfPath!);
  console.log(`    Total pages extracted: ${allPages.length}`);

  // Filter out the skip range (already in cia_criteria)
  const eligiblePages = allPages.filter(
    (p) => p.pageNumber < skipFrom || p.pageNumber > skipTo
  );
  console.log(
    `    Skipping pages ${skipFrom}–${skipTo} (already in cia_criteria)`
  );
  console.log(`    Pages to ingest: ${eligiblePages.length}`);

  const chunks = chunkPages(eligiblePages);
  console.log(`    Chunks created: ${chunks.length} (target ~${chunkWords} words each)\n`);

  if (clearFirst) {
    console.log("🗑️   --clear flag set. Deleting existing pdf_knowledge rows...");
    const { error } = await supabase.from("pdf_knowledge").delete().neq("id", 0);
    if (error) console.error("    Delete error:", error.message);
    else console.log("    ✓ Table cleared\n");
  }

  // Ingest chunks one by one
  let success = 0;
  let failed  = 0;

  for (let i = 0; i < chunks.length; i++) {
    const chunk = chunks[i];
    const progress = `[${i + 1}/${chunks.length}]`;

    try {
      const embedding = await embedText(chunk.content);

      const { error } = await supabase.from("pdf_knowledge").insert({
        content:    chunk.content,
        section:    chunk.section,
        page_start: chunk.pageStart,
        page_end:   chunk.pageEnd,
        embedding,
      });

      if (error) {
        console.error(`  ${progress} ✗  ${chunk.section} (p.${chunk.pageStart}): ${error.message}`);
        failed++;
      } else {
        console.log(`  ${progress} ✓  ${chunk.section} (p.${chunk.pageStart}–${chunk.pageEnd})`);
        success++;
      }
    } catch (err: any) {
      console.error(`  ${progress} ✗  Embedding error: ${err.message}`);
      failed++;
    }

    // Small delay to avoid rate limits
    await new Promise((res) => setTimeout(res, 120));
  }

  console.log(`\n✅  Done! ${success} chunks ingested, ${failed} failed.`);
  if (failed > 0) {
    console.log("    Re-run with --clear to retry from scratch, or check the errors above.");
  }
}

main().catch(console.error);

File: scripts/ingest-pdf-knowledge.py
"""
scripts/ingest-pdf-knowledge.py

Chunks and embeds the full KMS PDF into the pdf_knowledge Supabase table.
Skips nothing by default — indexes everything including pages 75-253
(the prose/Quranic explanations there are different from the bare criteria names
already in cia_criteria).

REQUIREMENTS:
  pip install pymupdf requests python-dotenv

RUN:
  python3 scripts/ingest-pdf-knowledge.py --pdf /path/to/FullQCB.pdf

OPTIONS:
  --pdf          <path>   Path to the PDF  [REQUIRED]
  --chunk-words  <n>      Target words per chunk  [default: 300]
  --clear                 Delete all existing rows before ingesting
"""

import argparse
import os
import sys
import time
import json
import re
import requests
import fitz  # pymupdf
from dotenv import load_dotenv

load_dotenv(".env.local")

# ── Config ────────────────────────────────────────────────────────────────────

OPENROUTER_API_KEY = os.getenv("OPENROUTER_API_KEY")
SUPABASE_URL       = os.getenv("NEXT_PUBLIC_SUPABASE_URL")
SUPABASE_KEY       = os.getenv("SUPABASE_SERVICE_ROLE_KEY") or os.getenv("NEXT_PUBLIC_SUPABASE_ANON_KEY")
EMBEDDING_MODEL    = "openai/text-embedding-3-small"

if not OPENROUTER_API_KEY:
    sys.exit("❌  OPENROUTER_API_KEY missing in .env.local")
if not SUPABASE_URL or not SUPABASE_KEY:
    sys.exit("❌  Supabase credentials missing in .env.local")

# ── Section map (from the PDF table of contents) ──────────────────────────────

SECTIONS = [
    (1,   3,   "Daftar Isi"),
    (4,   14,  "Puzzle Kesuksesan"),
    (15,  74,  "Dasar Teori QCB QMB QSB"),
    (75,  141, "Quran Character Building (QCB)"),
    (142, 217, "Quran Mental Building (QMB)"),
    (218, 253, "Quran Soft Skill Building (QSB)"),
    (254, 258, "Cara Menanamkan KMS Quranik"),
    (259, 271, "Simulasi: Berpisah dengan Orangtua"),
    (272, 284, "Simulasi: Tekanan Akademik"),
    (285, 296, "Simulasi: Konflik Sosial"),
    (297, 309, "Simulasi: Perubahan Lingkungan"),
    (310, 321, "Simulasi: Kegagalan atau Penolakan"),
    (322, 333, "Simulasi: Situasi Tidak Terduga"),
    (334, 344, "Simulasi: Tekanan Moral atau Etika"),
    (345, 356, "Simulasi: Medsos dan Perbandingan Diri"),
    (357, 368, "Simulasi: Pengambilan Keputusan"),
    (369, 380, "Simulasi: Kehilangan atau Duka"),
    (381, 392, "Simulasi: Tantangan Identitas Diri"),
    (393, 403, "Simulasi: Kompetisi dan Persaingan"),
    (404, 419, "Simulasi: Kritik atau Teguran"),
    (420, 436, "Simulasi: Penolakan"),
    (437, 454, "Simulasi: Permintaan Bantuan"),
    (455, 474, "Simulasi: Tidak Adil atau Diskriminatif"),
    (475, 493, "Simulasi: Situasi Menyebalkan"),
    (494, 512, "Simulasi: Dominan atau Terlalu Mengatur"),
    (513, 528, "Simulasi: Pasif atau Tidak Kooperatif"),
    (529, 542, "Simulasi: Meremehkan atau Merendahkan"),
    (543, 558, "Simulasi: Emosional atau Marah"),
    (559, 570, "Simulasi: Manipulatif atau Memperalat"),
    (571, 587, "Simulasi: Tidak Konsisten"),
    (588, 596, "Simulasi: Kompetitif Berlebihan"),
    (597, 597, "Catatan Penutup"),
]

def get_section(page_number: int) -> str:
    for start, end, name in SECTIONS:
        if start <= page_number <= end:
            return name
    return f"Halaman {page_number}"

# ── PDF extraction ─────────────────────────────────────────────────────────────

def extract_pages(pdf_path: str) -> list[dict]:
    doc = fitz.open(pdf_path)
    pages = []
    for i, page in enumerate(doc):
        text = page.get_text("text").replace("\n", " ").strip()
        text = re.sub(r"\s+", " ", text)
        pages.append({"page": i + 1, "text": text})
    print(f"    Extracted {len(pages)} pages from PDF")
    return pages

# ── Chunking ───────────────────────────────────────────────────────────────────

def chunk_pages(pages: list[dict], chunk_words: int) -> list[dict]:
    # Group pages by section
    groups: dict[str, list[dict]] = {}
    order: list[str] = []

    for p in pages:
        section = get_section(p["page"])
        if section not in groups:
            groups[section] = []
            order.append(section)
        groups[section].append(p)

    chunks = []
    for section in order:
        group = groups[section]
        full_text = " ".join(p["text"] for p in group)
        full_text = re.sub(r"\s+", " ", full_text).strip()
        if not full_text:
            continue

        words = full_text.split()
        start = 0
        while start < len(words):
            end = min(start + chunk_words, len(words))
            chunk_text = " ".join(words[start:end]).strip()

            if len(chunk_text) > 80:  # skip near-empty chunks
                chunks.append({
                    "content": f"[{section}]\n{chunk_text}",
                    "section": section,
                    "page_start": group[0]["page"],
                    "page_end": group[-1]["page"],
                })
            start = end

    return chunks

# ── Embedding ──────────────────────────────────────────────────────────────────

def embed_text(text: str) -> list[float]:
    response = requests.post(
        "https://openrouter.ai/api/v1/embeddings",
        headers={
            "Authorization": f"Bearer {OPENROUTER_API_KEY}",
            "Content-Type": "application/json",
        },
        json={"model": EMBEDDING_MODEL, "input": text},
        timeout=30,
    )
    response.raise_for_status()
    return response.json()["data"][0]["embedding"]

# ── Supabase helpers ───────────────────────────────────────────────────────────

def supabase_delete_all():
    response = requests.delete(
        f"{SUPABASE_URL}/rest/v1/pdf_knowledge?id=gt.0",
        headers={
            "apikey": SUPABASE_KEY,
            "Authorization": f"Bearer {SUPABASE_KEY}",
        },
        timeout=15,
    )
    response.raise_for_status()

def supabase_insert(row: dict):
    response = requests.post(
        f"{SUPABASE_URL}/rest/v1/pdf_knowledge",
        headers={
            "apikey": SUPABASE_KEY,
            "Authorization": f"Bearer {SUPABASE_KEY}",
            "Content-Type": "application/json",
            "Prefer": "return=minimal",
        },
        json=row,
        timeout=30,
    )
    response.raise_for_status()

# ── Main ───────────────────────────────────────────────────────────────────────

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--pdf", required=True, help="Path to the full PDF")
    parser.add_argument("--chunk-words", type=int, default=300)
    parser.add_argument("--clear", action="store_true")
    args = parser.parse_args()

    print(f"\n📖  Reading PDF: {args.pdf}")
    pages = extract_pages(args.pdf)
    chunks = chunk_pages(pages, args.chunk_words)
    print(f"    Chunks created: {len(chunks)} (~{args.chunk_words} words each)\n")

    if args.clear:
        print("🗑️   --clear flag set. Deleting existing rows...")
        supabase_delete_all()
        print("    ✓ Table cleared\n")

    success = 0
    failed = 0

    for i, chunk in enumerate(chunks):
        label = f"[{i+1}/{len(chunks)}]"
        try:
            embedding = embed_text(chunk["content"])
            supabase_insert({
                "content":    chunk["content"],
                "section":    chunk["section"],
                "page_start": chunk["page_start"],
                "page_end":   chunk["page_end"],
                "embedding":  embedding,
            })
            print(f"  {label} ✓  {chunk['section']} (p.{chunk['page_start']}–{chunk['page_end']})")
            success += 1
        except Exception as e:
            print(f"  {label} ✗  {chunk['section']}: {e}")
            failed += 1

        time.sleep(0.12)  # avoid rate limits

    print(f"\n✅  Done! {success} chunks ingested, {failed} failed.")
    if failed:
        print("    Re-run with --clear to retry from scratch.")

if __name__ == "__main__":
    main()
